ccm_web/server/src/canvas/canvas.controller.ts

```diff
@@ -0,0 +1,64 @@
+import { Request, Response } from 'express'
+import {
+  Controller, Get, InternalServerErrorException, Query, Req, Res, UnauthorizedException
+} from '@nestjs/common'
+
+import { OAuthResponseQuery } from './canvas.interfaces'
+import { CanvasService } from './canvas.service'
+
+import baseLogger from '../logger'
+
+const logger = baseLogger.child({ filePath: __filename })
+
+@Controller('canvas')
+export class CanvasController {
+  constructor (private readonly canvasService: CanvasService) {}
+
+  @Get('redirectOAuth')
+  async redirectToOAuth (
+    @Req() req: Request, @Res() res: Response
+  ): Promise<void> {
+    logger.debug('Pulling session data, checking for Canvas token, and redirecting to Canvas for OAuth...')
+
+    if (req.session.data === undefined) {
+      throw new InternalServerErrorException('Session data is not available!')
+    }
+
+    const { ltiKey, userLoginId } = req.session.data
+
+    const token = await this.canvasService.findToken(userLoginId)
+    if (token !== null) {
+      logger.debug(`User ${userLoginId} already has a CanvasToken record, redirecting to home page...`)
+      return res.redirect(`/?ltik=${ltiKey}`)
+    }
+
+    const fullURL = `${this.canvasService.getAuthURL()}&state=${req.sessionID}`
+    logger.debug(`Full redirect URL: ${fullURL}`)
+    res.redirect(fullURL)
+  }
+
+  /*
+  Not behind ltijs authentication; this seems necessary, right?
+  See provider.whitelist in LTIService
+  */
+  @Get('returnFromOAuth')
+  async returnFromOAuth (
+    @Query() query: OAuthResponseQuery, @Req() req: Request, @Res() res: Response
+  ): Promise<void> {
+    logger.debug('Comparing session to state parameter, and creating new Canvas token if matching')
+    logger.debug(`Session ID: ${req.sessionID}`)
+    logger.debug(JSON.stringify(req.session, null, 2))
+
+    if (req.sessionID !== query.state) throw new UnauthorizedException()
```

Comment:
So for now, I'm just passing the `sessionID` as the state variable, and comparing when it comes back. Does this seem secure? We could also use the NodeJS `crypto` library to create another secret and store that in the session. Not sure if that's more secure or not.

Reply:
```
The crypto module is mostly useful as a tool for implementing cryptographic protocols such as TLS and https. For most users, the built-in tls module and https module should more than suffice. However, for the user that only wants to use small parts of what's needed for full-scale cryptography or is crazy/desperate enough to implement a protocol using OpenSSL and Node.js: Read on.
```
Based on my reading at this [website](https://nodejs.org/en/knowledge/cryptography/how-to-use-crypto-module/) I don't think we need any `crypto` library for our use case.

I would like to add a `log.warn` to monitor this case in prod log if at all this happens.

Reply:
I guess my question is, is the session ID something someone could fabricate easier than some other secret we generate in the backend and store inside the session? We don't have to use `crypto` to generate the secret, but that seems like one option.

I added a log message in this commit: https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/92/commits/d8922b65510b0458aa7633283cf8c1e3c72ae828